README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# sclr

<!-- badges: start -->

[![CRAN
status](https://www.r-pkg.org/badges/version/sclr)](https://cran.r-project.org/package=sclr)
[![Travis build
status](https://travis-ci.org/khvorov45/sclr.svg?branch=master)](https://travis-ci.org/khvorov45/sclr)
[![codecov](https://codecov.io/gh/khvorov45/sclr/branch/master/graph/badge.svg)](https://codecov.io/gh/khvorov45/sclr)
<!-- badges: end -->

The goal of sclr is to fit the scaled logit model from Dunning (2006)
using the maximum likelihood method.

## Installation

Install the [CRAN](https://CRAN.R-project.org) version with

``` r
install.packages("sclr")
```

Or the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("khvorov45/sclr")
```

## Model

The model is logistic regression with an added parameter for the top
asymptote. For model specification, log likelihood, scores and second
derivatives see `vignette("sclr-math")`. Documentation of the main
fitting function `?sclr` has details on how the model is fit.

## Example

Usage is similar to other model fitting functions like `lm`.

``` r
library(sclr)
fit <- sclr(status ~ logHI, one_titre_data) # included simulated data
summary(fit)
#> Call: status ~ logHI
#> 
#> Parameter estimates
#>       theta      beta_0  beta_logHI 
#> -0.03497876 -5.42535734  2.14877741 
#> 
#> 95% confidence intervals
#>                 2.5 %      97.5 %
#> theta      -0.1350572  0.06509969
#> beta_0     -6.4417802 -4.40893449
#> beta_logHI  1.8146909  2.48286390
#> 
#> Log likelihood: -2469.765
```

For more details see `vignette("sclr-usage")`.

## References

Dunning AJ (2006). “A model for immunological correlates of protection.”
Statistics in Medicine, 25(9), 1485-1497. doi: 10.1002/sim.2282.


NEWS.md
# sclr 0.3.0

- Reparameterised the model so that all of the parameters are unconstrained.
New baseline is the logit transformation of the old baseline.

- Added the gradient ascent algorithm to handle cases with high baseline.

- Added a warning for a possible baseline of 1.

- Added the ability to check for a possible baseline of 1 with `check_baseline`.

- Added `logLik` method to access likelihood from the fit object.

- Added a warning message when the model is fit with no covariates.

# sclr 0.2.0

- Added `sclr_ideal_data` function to simulate ideal data for the model.

- Made simulations in data-raw self-contained.

- Added the ability to return parameter names that are more conventional (e.g.
"(Intercept)" instead of "beta_0"). See `conventional_names` argument in
`?sclr`.

- Made convergence stricter to avoid local maxima. Argument `n_conv` to `sclr`
and `sclr_fit` sets the number of times the algorithm has to converge. Best set
(the one with maximum likelihood) is chosen out of `n_conv` sets. Previously,
the algorithm only converged once.

- `sclr_log_likelihood` can now be called with a model matrix and a model response.

- Minor performance optimisations.

# sclr 0.1.0

First release.

## Main features

- Fits the scaled logit model using the Newton-Raphson method.

- Supports the predict method for the expected value of the linear beta X part
of the model.

- Can look for covariate values corresponding to a particular protection level.

R/sclr.R
# sclr class functions
# Arseniy Khvorov
# Created 2019/07/31
# Last edit 2019/11/04

#' Fits the scaled logit model
#'
#' Used to fit the scaled logit model from Dunning (2006).
#'
#' The model is logistic regression with an added parameter for the top
#' asymptote. That parameter is reported as \code{theta} (or \code{(Baseline)}
#' if \code{conventional_names = TRUE}). Note that it is reported on the logit
#' scale. See \code{vignette("sclr-math")} for model specification,
#' log-likelihood, scores and second derivatives. The main default optimisation
#' algorithm is Newton-Raphson. Gradient ascent is used as a fallback by
#' default. Computing engine behind the fitting is \code{\link{sclr_fit}}.
#'
#' @param formula an object of class "formula": a symbolic description of the
#'   model to be fitted.
#' @param data a data frame.
#' @param ci_lvl Confidence interval level for the parameter estimates.
#' @inheritParams sclr_fit
#'
#' @return An object of class \code{sclr}. This is a list with the following
#'   elements:
#'
#'   \item{parameters}{Maximum likelihood estimates of the parameter values.}
#'
#'   \item{covariance_mat}{The variance-covariance matrix of the parameter
#'   estimates.}
#'   
#'   \item{algorithm}{Algorithm used.}
#'   
#'   \item{algorithm_return}{Everything the algorithm returned.}
#'
#'   \item{n_converge}{The number of Newton-Raphson iterations (including
#'   resets) that were required for convergence.}
#'
#'   \item{x}{Model matrix derived from \code{formula} and \code{data}.}
#'
#'   \item{y}{Response matrix derived from \code{formula} and \code{data}.}
#'
#'   \item{call}{The original call to \code{sclr}.}
#'
#'   \item{model}{Model frame object derived from \code{formula} and
#'   \code{data}.}
#'
#'   \item{terms}{Terms object derived from model frame.}
#'
#'   \item{ci}{Confidence intervals of the parameter estimates.}
#'
#'   \item{log_likelihood}{Value of log-likelihood calculated at the ML
#'   estimates of parameters.}
#'   
#'   \item{formula}{Passed formula.}
#'   
#'   \item{data}{Passed data.}
#'
#'   Methods supported: \code{\link[=print.sclr]{print}},
#'   \code{\link[=vcov.sclr]{vcov}}, \code{\link[=coef.sclr]{coef}},
#'   \code{\link[=model.frame.sclr]{model.frame}},
#'   \code{\link[=model.matrix.sclr]{model.matrix}},
#'   \code{\link[=summary.sclr]{summary}}, \code{\link[=predict.sclr]{predict}},
#'   \code{\link[=tidy.sclr]{tidy}} (\code{\link{broom}} package),
#'   \code{\link[=logLik.sclr]{logLik}}.
#'
#' @references Dunning AJ (2006). "A model for immunological correlates of
#'   protection." Statistics in Medicine, 25(9), 1485-1497.
#'   \url{https://doi.org/10.1002/sim.2282}.
#'
#' @examples
#' library(sclr)
#' fit1 <- sclr(status ~ logHI, one_titre_data)
#' summary(fit1)
#' @importFrom stats confint model.frame model.matrix model.response
#' @importFrom rlang abort warn
#'
#' @export
sclr <- function(formula, data = NULL, 
                 ci_lvl = 0.95, 
                 tol = 10^(-7),
                 algorithm = c("newton-raphson", "gradient-ascent"),
                 nr_iter = 2e3, ga_iter = 2e3, n_conv = 3,
                 conventional_names = FALSE, seed = NULL) {
  
  if (!inherits(formula, "formula") || missing(formula)) 
    abort("must supply a formula")

  cl <- match.call()
  mf <- model.frame(formula, data, drop.unused.levels = TRUE)

  # Design matirix
  mt <- attr(mf, "terms")
  x <- model.matrix(mt, mf)
  if (ncol(x) == 1) warn("no unique solution with no covariates")

  # Response vector
  y <- model.response(mf)
  if (!all(y %in% c(0, 1))) abort("response should be a vector with 0 and 1")
  if (!is.numeric(y)) abort("response should be numeric")

  # Actual model fit
  fit <- sclr_fit(
    y, x, tol, algorithm, nr_iter, ga_iter, n_conv, conventional_names, seed
  )

  # Build the return list
  fit <- new_sclr(fit, x, y, cl, mf, mt)
  fit$ci <- confint(fit, level = ci_lvl)
  fit$log_likelihood <- sclr_log_likelihood(fit)
  fit$formula <- formula
  fit$data <- data
  fit
}

#' Create a new \code{sclr} object
#' 
#' \code{new_sclr} creates the object \code{\link{sclr}} returns.
#' \code{is_sclr} checks if the object is of class \code{sclr}.
#'
#' @param fit A list returned by \code{\link{sclr_fit}}.
#' @param x Model matrix.
#' @param y Model response.
#' @param cl Call.
#' @param mf Model frame.
#' @param mt Model terms.
#'
#' @return \code{sclr} object
#' @export
new_sclr <- function(fit, x, y, cl, mf, mt) {
  stopifnot(is.list(fit))
  stopifnot(is.matrix(x))
  stopifnot(is.numeric(y))
  stopifnot(is.call(cl))
  stopifnot(is.data.frame(mf))
  stopifnot(inherits(mt, "terms"))
  fit <- c(fit, list(
    x = x, y = y,
    call = cl, model = mf, terms = mt
  ))
  class(fit) <- "sclr"
  fit
}

#' @rdname new_sclr
#' @export
is_sclr <- function(fit) "sclr" %in% class(fit)
